python/lsst/resources/s3.py

```diff
@@ -116,7 +122,50 @@ def _translate_client_error(err: ClientError) -> None:
 
 
 class S3ResourcePath(ResourcePath):
-    """S3 URI resource path implementation class."""
+    """S3 URI resource path implementation class.
+
+    Notes
+    -----
+    In order to configure the behavior of instances of this class, the
+    environment variable is inspected:
+
+    - LSST_S3_USE_THREADS: May be True, False, or None. Sets whether threading
+    is used for downloads, with a value of None defaulting to boto's default
+    value. Users may wish to set it to False when the downloads will be started
+    within threads other than python's main thread.
+    """
+
+    use_threads: bool | None = None
+    """Explicitly turn on or off threading in use of boto's download_fileobj.
+    Setting this to None results in boto's default behavior."""
+
+    @property
+    def _transfer_config(self) -> TransferConfig:
+        if self.use_threads is None:
```

Comment:
Doesn't this block get executed every single time by default? Even if the environment variable parsing hasn't changed and the environment variable itself is the same? How about:

```python
_read_use_threads_environ = False
_cached_use_threads = None
def _read_use_threads_from_environment() -> bool:
    if _read_use_threads_environ:
        return _cached_use_threads
    try:
        s3_use_threads = os.environ[...]
    except KeyError:
        s3_use_threads = None
    if s3_use_threads is None:
        ...
    elif ...:
        ...

    _read_use_threads_environ = True
    _cached_use_threads = use_threads
    return use_threads
```
and then call this from `_transfer_config` ? It seems like the environment parsing is no longer a quick thing because you want to support a myriad of different values rather than 0 1 or unset.

Reply:
How about:
```

    @cached_property
    def _environ_use_threads(self):
        try:
            use_threads_str = os.environ["LSST_S3_USE_THREADS"]
        except KeyError:
            use_threads_str = None

        if use_threads_str is None:
            use_threads = None
        elif use_threads_str.lower() in ["t", "true", "yes", "y", "1"]:
            use_threads = True
        elif use_threads_str.lower() in ["f", "false", "no", "n", "0"]:
            use_threads = False
        elif use_threads_str.lower() in ["none", ""]:
            use_threads = None
        else:
            raise ValueError(
                f'LSST_S3_USE_THREADS value of "{use_threads_str}" is not True, False, or None.'
            )

        return use_threads

    @property
    def _transfer_config(self) -> TransferConfig:
        if self.use_threads is None:
            self.use_threads = self._environ_use_threads

        if self.use_threads is None:
            transfer_config = TransferConfig()
        else:
            transfer_config = TransferConfig(use_threads=self.use_threads)

        return transfer_config
```

Reply:
Works for me. I keep forgetting about cached_property.